src/shared/bpf-object.c

```diff
@@ -0,0 +1,141 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/resource.h>
+
+#include "bpf-object.h"
+#include "bpf-program-v2.h"
+#include "fd-util.h"
+#include "set.h"
+
+int bpf_object_new(const unsigned char *mem_buf, size_t size, struct bpf_object **object) {
+        _cleanup_(bpf_object_freep) struct bpf_object *p = NULL;
+        /* Standard error code. */
+        long err;
+
+        assert(object);
+
+        DECLARE_LIBBPF_OPTS(
+                        bpf_object_open_opts,
+                        opts);
+
+        p = bpf_object__open_mem(mem_buf, size, &opts);
+        /* Should either point to a valid BPF object or store standard error code. */
+        if (!p)
+                return -EINVAL;
+
+        err = libbpf_get_error(p);
+        if (err)
+                return -err;
+
+        *object = TAKE_PTR(p);
+        return 0;
+}
+
+struct bpf_object *bpf_object_free(struct bpf_object *object) {
+        if (object)
+                bpf_object__close(object);
+
+        return NULL;
+}
+
+int bpf_object_load(struct bpf_object *object) {
+        assert(object);
+
+        /* libbpf logs kernel warnings with pr_warn. */
+        if (bpf_object__load(object))
+                return -EINVAL;
+
+        return 0;
+}
+
+ DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(prog_hash_ops, void, trivial_hash_func, trivial_compare_func, BPFProgramV2, bpf_program_v2_free);
+
+int bpf_object_get_programs(const struct bpf_object *object, Set **progs) {
```

Comment:
return parameters are usually prefixed with ret_ - eg, this would be ret_progs. Then you can rename _progs -> progs (same in the function above)